controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -118,9 +118,31 @@ else if (retryBecauseNewFailure(application, report.jobType()))
     /** Returns whether all production zones listed in deployment spec last were successful on the currently deploying change. */
     private boolean deploymentComplete(LockedApplication application) {
         if ( ! application.deploying().isPresent()) return true;
-        return order.jobsFrom(application.deploymentSpec()).stream()
-                .filter(JobType::isProduction)
-                .allMatch(jobType -> application.deploymentJobs().isSuccessful(application.deploying().get(), jobType));
+        Change change = application.deploying().get();
+
+        for (JobType job : order.jobsFrom(application.deploymentSpec())) {
+            if ( ! job.isProduction()) continue;
+
+            Optional<ZoneId> zone = job.zone(this.controller.system());
+            if ( ! zone.isPresent()) continue;
```

Comment:
Is it possible to have production jobs with no zone?

Reply:
Those aren't separate jobs. The point of this optional is that not all systems have all zones.

Reply:
The jobs here come from a valid deployment spec, so all zones listed should exist?

Reply:
Not in CD or whatever. I didn't decide that job.zone(system) should return Optional, but it does ...

Reply:
ApplicationApiHandler.java#795. 
I guess you can fake it in unit tests, though.

Reply:
I don't see what you mean. I actually need to handle the Optional. Anyway, this isn't important ...